[pwrmgr] Add documentation on a few countermeasures (#11455)
lowRISC/opentitan: hw/ip/pwrmgr/doc/_index.md

The slow FSM requests the AST to power up the main domain and high speed clocks.
Once those steps are done, it requests the [fast FSM](#fast-clock-domain-fsm) to begin operation.
The slow FSM also handles power isolation controls as part of this process.

Once the fast FSM acknowledges the power-up completion, the slow FSM transitions to `Idle` and waits for a power down request.
When a power down request is received, the slow FSM turns off AST clocks and power as directed by software configuration.
This means the clocks and power are not always turned off, but are rather controlled by software configurations in {{< regref "CONTROL" >}} prior to low power entry .
Once these steps are complete, the slow FSM transitions to a low power state and awaits a wake request, which can come either as an actual wakeup, or a reset event (for example always on watchdog expiration).

#### Sparse FSM

Since the slow FSM is sparsely encoded, it is possible for the FSM to end up in an undefined state if attacked.
When this occurs, the slow FSM sends an `invalid` indication to the fast FSM and forcibly powers off and clamps everything.

> The clocks are kept on however to allow the fast FSM to operate if it able to receive the `invalid` indication.

Reviewer: if it able to => if it is able to

Author: fixed!

For reset requests, fall through and aborts are not checked and the system simply resets directly.
Note in this scenario the slow FSM is not requested to take over.

#### Sparse FSM

Since the fast FSM is sparsely encoded, it is possible for the FSM to end up in an undefined state if attacked.

> When this occurs, the fast FSM forcibly disables all clocks and holds the system in reset.

Reviewer 2: ```suggestion
When this occurs, the fast FSM forcibly disables all clocks and requests a system reset.
```

Requiring a POR is a bit undesirable... any other options?

Author: yeah this one actually just kills the system in place.
Mostly because if the fsm gets out of wack I cannot guarantee correct synchronization anymore between the slow/fast FSMs.  So it's not easy to recover (not impossible, would require more code).

Reviewer 2: Hmm, imagine this protocol: if slow fsm gets out of whack it triggers a por-like reset, which gets or'ed into the por / scan reset. If the fast fsm gets out of whack sends a request for slow fsm to do the same. Request for enhancement perhaps? Well, the question is whether the current solution could somehow trigger a POR, or will the chip just sit there until the power goes off and then on again spontaneously?

Author: it's actually not that trivial to trigger a POR. If you let digital logic trigger POR for example, you kind of have a chicken and egg problem.  The digital logic is in an undefined state until a POR arrives (in the powering up case), so if you allow digital logic to directly affect the PR, you can get stuck in a situation where the POR does not properly assert.  

So to properly do this the AST actually needs to have a separation between the real POR and what it considers a POR request (which is handled slightly differently).  AST does have a POR input right now for this, but we only agreed on tying it to a pad input and not expose it other digital logic.  I would be a bit weary of exposing that at this point. 

Coming back to whether this is worth enhancing.  My view is that damaging power management behavior is very very destructive. So in this scenario, the device effectively stops working and requires outside intervention.  Outside intervention can be done in a couple ways

1. manually cycle the external power to create a real POR
2. assert the external POR pin to create a POR request and hope that AST responds to it.

Reviewer 2: What you are saying would mean scan_rst_ni is just as dangerous as the proposed change, since scan reset acts like POR.

My problem with the current behavior is that the chip will become totally unresponsive, and we would get no indication of why... I guess this reinforces the popular rule that if something seems to be wedged you need to do a power cycle, but that can be really annoying, and when chips are deployed in remote locations (like inside a data center) they can be very difficult to handle.

Author: no that's not correct. scan_rst_ni does not have the same behavior as the "real" POR inside AST.

Yes, but again remember this behavior surfaces from the power manager going into an unknown state, which is a very intrusive attack. In some sense you probably want this to halt because it would imply something is very wrong in the system. 

This is also why however we still have a POR input, so an accompany host that notices the system is stuck can decide to rescue it.

Author: Having said all that, while we cannot directly trigger a real POR, we could do something like reset the state machines to their default states (on both sides), and try to re-synchronize the entire behavior.

I'm just honestly not sure it's worth doing for power manager due to the implications it has on the system.
